github.com/cockroachdb/pebble/internal/metamorphic: TestMeta failed

github.com/cockroachdb/pebble/internal/metamorphic.TestMeta [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/19770363?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/19770363?buildTab=artifacts#meta) on refs/heads/master @ [abf8da2b6b97](https://github.com/cockroachdb/pebble/commits/abf8da2b6b97):


```
  table_cache_shards=8
  validate_on_ingest=false
  wal_dir=data/wal
  wal_bytes_per_sync=0
  secondary_cache_size_bytes=0
  create_on_shared=0

[Value Separation]
  enabled=true
  minimum_size=11
  max_blob_reference_depth=8

[WAL Failover]
  secondary_dir=data/wal_secondary
  primary_dir_probe_interval=173.223µs
  healthy_probe_latency_threshold=4.319103ms
  healthy_interval=13.940766ms
  unhealthy_sampling_interval=375ns
  unhealthy_operation_latency_threshold=2.965µs
  elevated_write_stall_threshold_lag=15.466016ms

[Level "0"]
  block_restart_interval=34
  block_size=16
  block_size_threshold=59
  compression=NoCompression
  filter_policy=rocksdb.BuiltinBloomFilter
  filter_type=table
  index_block_size=8192
  target_file_size=1024

[TestOptions]
  key_format=testkeys
  ingest_using_apply=true
  threads=8
  disable_block_property_collector=true
  disable_value_blocks_for_ingest_sstables=true
  async_apply_to_db=true
  external_storage_enabled=true
  seed_efos=7644404428757735908
  io_latency_mean=3.833033ms
  io_latency_probability=0.0064562348
  io_latency_seed=1004241687695870612
  use_excise=true
  disable_downloads=true
  use_jemalloc_size_classes=true

===== OPS =====
/artifacts/meta/250524-084354.6254171497873/ops
To reduce:  go test ./internal/metamorphic -tags invariants -run 'TestMeta$' --compare "/artifacts/meta/250524-084354.6254171497873/{standard-000,random-023}" --try-to-reduce -v
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run 'TestMeta$' ./internal/metamorphic -seed 1748076234625214747 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMeta.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: PEBBLE-454

I spent quite a few hours debugging this. The filenum, blocknum and block offset that we are retrieving are correct. At the level of `cachedReader`, we are returning the correct value (`fkpiuvrzxtxddbaug`):
```
file 71, block 1, offset 0, len 17: "fkpiuvrzxtxddbaug"
goroutine 9 [running]:
runtime/debug.Stack()
	/Users/radu/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.3.darwin-arm64/src/runtime/debug/stack.go:26 +0x64
runtime/debug.PrintStack()
	/Users/radu/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.3.darwin-arm64/src/runtime/debug/stack.go:18 +0x1c
github.com/cockroachdb/pebble/sstable/blob.(*cachedReader).GetUnsafeValue(0x140005303d0, {0x102e0e138, 0x103310f40}, {0x1025ecd0c?, 0x843a28?, 0x140?, 0x25f1cd8?}, {0x140005307e0, 0x1400021e740, 0x0, ...})
	/Users/radu/go/src/github.com/cockroachdb/pebble/sstable/blob/fetcher.go:212 +0x35c
github.com/cockroachdb/pebble/sstable/blob.(*ValueFetcher).retrieve(0x135e08a2e?, {0x102e0e138?, 0x103310f40?}, {0x1026078dc?, 0x40?, 0x0?, 0x2cd0f60?})
	/Users/radu/go/src/github.com/cockroachdb/pebble/sstable/blob/fetcher.go:139 +0x214
github.com/cockroachdb/pebble/sstable/blob.(*ValueFetcher).Fetch(0x140005302c0, {0x102e0e138, 0x103310f40}, {0x135e08a2e?, 0x1032d3a80?, 0x102ddd480?}, 0x47, 0x11, {0x14000843b28?, 0x1025eec7c?, ...})
	/Users/radu/go/src/github.com/cockroachdb/pebble/sstable/blob/fetcher.go:96 +0x64
github.com/cockroachdb/pebble/internal/base.(*LazyValue).fetchValue(0x14000843b88?, {0x102e0e138?, 0x103310f40?}, {0x0?, 0xf?, 0x14000843c68?})
	/Users/radu/go/src/github.com/cockroachdb/pebble/internal/base/lazy_value.go:222 +0x7c
github.com/cockroachdb/pebble/internal/base.(*LazyValue).Value(0x0?, {0x0?, 0x1400058a0d0?, 0x0?})
	/Users/radu/go/src/github.com/cockroachdb/pebble/internal/base/lazy_value.go:212 +0x44
github.com/cockroachdb/pebble/internal/base.(*InternalValue).Value(...)
	/Users/radu/go/src/github.com/cockroachdb/pebble/internal/base/value.go:79
github.com/cockroachdb/pebble.(*Iterator).ValueAndErr(0x14000530008)
	/Users/radu/go/src/github.com/cockroachdb/pebble/iterator.go:2291 +0x30
github.com/cockroachdb/pebble.(*Iterator).Value(...)
	/Users/radu/go/src/github.com/cockroachdb/pebble/iterator.go:2281
github.com/cockroachdb/pebble/metamorphic.(*retryableIter).Value(...)
	/Users/radu/go/src/github.com/cockroachdb/pebble/metamorphic/retryable.go:188
github.com/cockroachdb/pebble/metamorphic.iteratorPos({{0x102b6d9a4, 0x8}, 0x1032dc600, 0x140001833b0, {0x1032cb990, 0x1, 0x1}, 0x102e006c8, 0x102e006d0, 0x0, ...}, ...)
	/Users/radu/go/src/github.com/cockroachdb/pebble/metamorphic/ops.go:1464 +0x154
github.com/cockroachdb/pebble/metamorphic.(*iterNextOp).run(0x140003b4420, 0x1400016dc20, {0x140003b45d0?, 0x1400061e0c0?, 0x0?})
```

This value is cached inside `LazyValue`. Other iterator operations happen (which likely do more stuff with the same underlying `cachedReader`). At the problematic SeekGE operation, we return the value without fetching it so we still had the LazyValue somewhere in the iterator stack. But the cached value is a pointer inside a buffer that was released by now.

```
iter29.Next("") // [true,"sxnozdbbktx@4","fkpiuvrzxtxddbaug",["sxnozdbbktx@4","thjuzcgdtk@11")=>{"@4"="rntvvyxvjbfnpgeivd"}*] <nil> #2527
iter29.First() // [true,"sxnozdbbktx@4","fkpiuvrzxtxddbaug",["sxnozdbbktx@4","thjuzcgdtk@11")=>{"@4"="rntvvyxvjbfnpgeivd"}] <nil> #2529
iter30.Next("kchq@18") // [invalid] cannot use limit with prefix iteration #2530
iter29.First() // [true,"sxnozdbbktx@4","fkpiuvrzxtxddbaug",["sxnozdbbktx@4","thjuzcgdtk@11")=>{"@4"="rntvvyxvjbfnpgeivd"}] <nil> #2531
iter29.SeekPrefixGE("opyymrmrepa") // [false] pebble: SeekPrefixGE supplied with key outside of lower bound #2535
iter29.First() // [true,"sxnozdbbktx@4","fkpiuvrzxtxddbaug",["sxnozdbbktx@4","thjuzcgdtk@11")=>{"@4"="rntvvyxvjbfnpgeivd"}*] <nil> #2536
iter29.Last() // [true,"thjuzcgdtk","mmerysoevmhd",["sxnozdbbktx@4","thjuzcgdtk@11")=>{"@4"="rntvvyxvjbfnpgeivd"}] <nil> #2540
iter29.InternalNext() // switching from reverse to forward via internalNext is prohibited #2541
panic: sxnozdbbktx@4 = "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00": {1 0} blob file num 000071 fetcher: *blob.ValueFetcher(0x140005302c0)
```

I'm not sure what the correct fix is. Maybe we should stop caching the value in LazyValue altogether?

CC @jbowens

--- a/internal/base/lazy_value.go
+++ b/internal/base/lazy_value.go
@@ -157,20 +157,15 @@ type LazyValue struct {
 
 // LazyFetcher supports fetching a lazy value.
 //
-// Fetcher and Attribute are to be initialized at creation time. The fields
-// are arranged to reduce the sizeof this struct.
+// The fields are to be initialized at creation time.
 type LazyFetcher struct {
 	// Fetcher, given a handle, returns the value.
 	Fetcher ValueFetcher
-	err     error
-	value   []byte
 	// Attribute includes the short attribute and value length.
 	Attribute AttributeAndLen
 	// BlobFileNum identifies the blob file containing the value. It is only
 	// populated if the value is stored in a blob file.
 	BlobFileNum DiskFileNum
-	fetched     bool
-	callerOwned bool
 }
 
 // ValueFetcher is an interface for fetching a value.
@@ -193,32 +188,12 @@ type ValueFetcher interface {
 
 // Value returns the underlying value.
 func (lv *LazyValue) Value(buf []byte) (val []byte, callerOwned bool, err error) {
-	if lv.Fetcher == nil {
+	f := lv.Fetcher
+	if f == nil {
 		return lv.ValueOrHandle, false, nil
 	}
-	// Do the rest of the work in a separate method to attempt mid-stack
-	// inlining of Value(). Unfortunately, this still does not inline since the
-	// cost of 85 exceeds the budget of 80.
-	//
-	// TODO(sumeer): Packing the return values into a struct{[]byte error bool}
-	// causes it to be below the budget. Consider this if we need to recover
-	// more performance. I suspect that inlining this only matters in
-	// micro-benchmarks, and in actual use cases in CockroachDB it will not
-	// matter because there is substantial work done with a fetched value.
-	return lv.fetchValue(context.TODO(), buf)
-}
-
-// INVARIANT: lv.Fetcher != nil
-func (lv *LazyValue) fetchValue(
-	ctx context.Context, buf []byte,
-) (val []byte, callerOwned bool, err error) {
-	f := lv.Fetcher
-	if !f.fetched {
-		f.fetched = true
-		f.value, f.callerOwned, f.err = f.Fetcher.Fetch(ctx,
-			lv.ValueOrHandle, f.BlobFileNum, f.Attribute.ValueLen, buf)
-	}
-	return f.value, f.callerOwned, f.err
+	return f.Fetcher.Fetch(context.TODO(),
+		lv.ValueOrHandle, f.BlobFileNum, f.Attribute.ValueLen, buf)
 }
 
 // Len returns the length of the value.

--- a/internal/base/lazy_value_test.go
+++ b/internal/base/lazy_value_test.go
@@ -68,7 +68,7 @@ func TestLazyValue(t *testing.T) {
 		require.Equal(t, []byte("foo"), getValue(fooLV3, callerOwned))
 		require.Equal(t, 1, numCalls)
 		require.Equal(t, []byte("foo"), getValue(fooLV3, callerOwned))
-		require.Equal(t, 1, numCalls)
+		require.Equal(t, 2, numCalls)
 		require.Equal(t, 3, fooLV3.Len())
 		attr, hasAttr := fooLV3.TryGetShortAttribute()
 		require.True(t, hasAttr)

